### ISSUE TITLE
Queuing should not be the default behavior

### ISSUE THREAD

**Issue by khaledsobhy83** [dotnet-state-machine/stateless#223]
I was using stateless library before introducing the queuing part and everything was just fine with me. Now after upgrading to the latest .Net Standard version, the queuing of triggers cause my application to totally fail. The point is that previously when calling .Fire I was expecting the event to be triggered but now it will be queued if the parent is still being processed and will not be executed. Assume a scenario when having two states Processing and checkpointing every one of these states has a begin and finish triggers, the beginprocessing of the first state fires the begincheckpointing of the second state then fires its finishprocessing event, assuming that when begincheckpoint is fired, it will call the finishcheckpoint first then the finishprocessing will be triggered. With the introduction of queuing part what will happen is that begincheckpoint, finishprocessing will be added to the queue in the wrong order, the correct order should be begincheckpoint, finishcheckpoint then finishprocessing.

I think the queuing is a nice feature but it should not be the default behavior, there should be a way to enforce the execution when calling .Fire like before.

**Comment by HenningNT** [comment]
I tend to disagree. Run-to-completion is important to guarantee atomic execution of events and state transitions. Therefore, allowing Fire to execute while processing a trigger should not be allowed.

**Comment by khaledsobhy83** [comment]
Exactly, this is what I want, Run-to-Completion, but what is currently happening is that instead of running and executing the trigger, they will be queued which does not mean actual execution and using the queue will end up of wrong order of items in the queue. Again here is the scenario and please tell me how to handle it
private void BeginProcessRecords()
        {
            _stateMachine.Fire(Trigger.BeginCheckPoint); // This calls FinishCheckpoint
            _stateMachine.Fire(Trigger.FinishProcessRecords); // This cannot be triggered before FinishCheckPoint is triggered
        }

What will happen is that During the execution of BeginProcessRecords, the BeginCheckPoint and FinishProcessRecords will be queues which is completely wrong, I want FinishCheckPoint to be executed before FinishProcessRecords.
Can you tell me how to guarantee that FinishProcessRecords will be triggered after FinishCheckpoint?

**Comment by HenningNT** [comment]
We'll, I "solved" my problem by posting all events to an external queue. I don't call Fire at all onside the state machine, I have timer that empties the queue at regular intervals. This allows me to also post lifo if needed.

**Comment by khaledsobhy83** [comment]
I added another method that does not use the queue and did a commit. not sure how the review process works. It will be another optional way to fire events the old way before the queuing part.

**Comment by khaledsobhy83** [comment]
#225

**Comment by khaledsobhy83** [comment]
@HenningNT I do not think that having an external queue with a timer will solve my problem...I can prove it...This is a link to AWS KCL, it was developed using Stateless and when upgrading to your .Net core version, it completely failed. I Think your test cases need to be revisited.
KCL Link:
https://github.com/awslabs/amazon-kinesis-client-net/issues/8

Note: Because of this, no one is able to build .net core version of KCL. I think someone here needs to check that.

**Comment by victorkirov** [comment]
Any update on this? I need to build the Amazon KCL project for .NET Core as @khaledsobhy83 mentioned

**Comment by HenningNT** [comment]
I'm reluctant to change the default behaviour, as that might break the other peoples projects.

I'm looking back at the example:
```
private void BeginProcessRecords()
{
_stateMachine.Fire(Trigger.BeginCheckPoint); // This calls FinishCheckpoint
_stateMachine.Fire(Trigger.FinishProcessRecords); // This cannot be triggered before FinishCheckPoint is triggered
}
```

Would this work:

```
private void BeginProcessRecords()
{
_stateMachine.Fire(Trigger.BeginCheckPoint); // This calls FinishCheckpoint
}

private void FinishCheckpoint()
{
// Your code goes here...
_stateMachine.Fire(Trigger.FinishProcessRecords); // This cannot be triggered before FinishCheckPoint is triggered
}
```

Another thought, could you use the Async version, and Wait for the completion of the first Fire? I haven't tried it, it seems a bit odd to use FireAsync to serialize the execution :-D
```
private void BeginProcessRecords()
{
var task = _stateMachine.FireAsync(Trigger.BeginCheckPoint); // This calls FinishCheckpoint
task.Wait(); // Wait for completion...
_stateMachine.Fire(Trigger.FinishProcessRecords); // This cannot be triggered before FinishCheckPoint is triggered
}
```

**Comment by victorkirov** [comment]
I hope so :P

**Comment by HenningNT** [comment]
I had a look at the code, and designed this diagram based on it
![kcl](https://user-images.githubusercontent.com/13196280/37464220-fa6781a8-2857-11e8-8450-3f88e673edb9.png)

I think there might be a "race condition", since Trigger.FinishProcessRecords is already Fired in BeginProcessRecords() after the _processor is done with ProcessRecords(...). If Trigger.FinishProcessRecords  is also Fired from the outside then you can't guarantee what will happen.

**Comment by khaledsobhy83** [comment]
@HenningNT it is good that you draw this diagram. This was developed based on a previous version from the Stateless library before introducing the queuing feature and I tried hard to update the KCL code to work with the new Stateless behavior but there was always a case that fails. I am not asking you to change the default behavior but at least give us a way to avoid queuing the event. In many cases, I want to make sure the Fire is executed and completed before calling FinishProcessing for example. I invested a lot on this and I am telling you that the only way to upgrade the AWS KCL to .net core version either you provide addition method to avoid the queuing part or KCL team has to find an alternative for Stateless. 
One more thing, this diagram is not showing everything because there are cases where the checkpointing is triggered explicitly.

I have initiated a pull request in which I just added a new method that avoids the queue and this is not going to affect anyone, but you will help your existing consumers to move on.

I really hope that to be understood and taken a place soon. Believe me, you will be helping many :).

**Comment by khaledsobhy83** [comment]
The race condition in the above diagram is because of Stateless not because of KCL. Stateless are the ones who decided to introduce the queuing feature.

**Comment by nblumhardt** [comment]
Would be great to be able to address this. Adding inheritance/virtual methods could have a perf impact, and does make maintaining invariants harder unless the shared surface area is very small.

> Stateless are the ones who decided to introduce the queuing feature.

It's a tricky one, honestly - the original bug report that led to this was a really early one - #11 - and the problem that queuing works around is rather gnarly. It was a community request, and community support, that got this done originally, so please try to understand that we're in a rather awkward position here :-)

There will also be invariants that will be broken without queueing, that we have no alternative solution for maintaining. I.e. the `OnExit()` event for a state, for example, might be called with the machine in a completely different state than the one that's supposed to be exiting.

The PR (thanks @khaledsobhy83!) seems incomplete with only one `FireNoQueue()` overload - there is a range of additional overloads, e.g. with parameters, that would need to be added for there to be full parity (and not an API wart to carry along).

Looking at all the options, including virtual methods etc., the least impactful one is probably adding:

```csharp

**Comment by HenningNT** [comment]
... and in the Fire method we could maybe do:

        public void Fire(TTrigger trigger)
        {
            if (_useQueue)
                InternalFire(trigger, new object[0]);
            else
                InternalFireOne(trigger, new object[0]);
        }

I think that would work for most people, without adding more complexity that absolutely needed.

**Comment by nblumhardt** [comment]
Placing the check in `Fire()` would cause different behavior depending on whether or not parameters are passed - making the same check in `InternalFire()` would cover all of the `Fire()` overloads.

**Comment by HenningNT** [comment]
Yes, that would work better!
I'd prefer to use an enum instead of bool, as these can be a bit more descriptive:

    public enum FiringMode
    {
        Immediate,
        Queued
    }

In InternalFire we can then do...

     switch (_firingMode)
            {
                case FiringMode.Immediate:
                   InternalFireOne(trigger, args);
                    return;
                case FiringMode.Queued:
                    // stuff goes here
                    break;
                default:
                    throw new InvalidOperationException();
            }

**Comment by HenningNT** [comment]
@IanElsinga I need to implement the Async version as well. Then you can use the "immediate" fireing mode. That might help...

**Comment by HenningNT** [comment]
We have duplicate code since we have InternalFire and InternalFireAsync. I'd like to call the async version only to reduce the code duplication, but I can't do that and support NET v4.0...
